### PR TITLE
First pass at implementing the redirection feature

### PR DESCRIPTION
**Pull request by felipemontoya**
This PR adds a new middleware capable of filtering out and reacting to requests by regex matching the url path to a list of per-microsite defined rules.
If one of the rules matches the current requets path, the action will be taken

### CODE REVIEW

File: edunext_openedx_extensions/edunext/middleware.py [hunk 1]
```diff
@@ -78,3 +81,68 @@ def clear_cache(sender, instance, **kwargs):  # pylint: disable=unused-argument
...
+    REDIRECT_LOGGEDIN = "redirect_loggedin"
+    REDIRECT_LOGGEDOUT = "redirect_loggedout"
+
+    def process_request(self, request):
+        """
+        This middleware processes the path of every request and determines if there
+        is a configured action to take.
+        """
+
+        if microsite.has_override_value("EDNX_CUSTOM_PATH_REDIRECTS"):
```
Comment (jfavellar90) [inline edunext_openedx_extensions/edunext/middleware.py]: Better if we return if the setting was not overriden? and if it was, just continue the code below?

File: edunext_openedx_extensions/edunext/middleware.py [hunk 2]
```diff
@@ -78,3 +81,68 @@ def clear_cache(sender, instance, **kwargs):  # pylint: disable=unused-argument
...
+            redirects = microsite.get_value("EDNX_CUSTOM_PATH_REDIRECTS", {})
```
Comment (jfavellar90) [inline edunext_openedx_extensions/edunext/middleware.py]: is this setting already defined in one microsite? if not, how it looks like?

File: edunext_openedx_extensions/edunext/middleware.py [hunk 3]
```diff
@@ -78,3 +81,68 @@ def clear_cache(sender, instance, **kwargs):  # pylint: disable=unused-argument
...
+                    key = values
+
+                path = request.path_info
+                regex_path_match = re.compile(regex.format(
+                    COURSE_ID_PATTERN=settings.COURSE_ID_PATTERN,
+                    USERNAME_PATTERN=settings.USERNAME_PATTERN,
+                ))
+
+                if regex_path_match.match(path):
+                    if key == self.LOGIN_REQUIRED:
```
Comment (jfavellar90) [inline edunext_openedx_extensions/edunext/middleware.py]: The list of possbile keys can grow in a way we should worry about using a different strategy than if-cases?

**Comment by felipemontoya** [comment]
This PR also requires the platform code to add a new middleware
`edunext_openedx_extensions.edunext.middleware.PathRedirectionMiddleware` near the end.

**Comment by felipemontoya** [comment]
@jfavellar90 thanks for the review.
About your questions. Is this already defined for a microsite: not yet, but if you wanted your life to be hell, you could make:  
```
  "EDNX_CUSTOM_PATH_REDIRECTS":{
    "/dashboard":"not_found"
  },
```

Also, yes, the list could grow which would make that particular site slower. Impact should not be much as long as the feature is not overused.

**Comment by felipemontoya** [comment]
@jfavellar90 any last thoughs?

**Review by jfavellar90 — APPROVED** [review]

**Comment by jfavellar90** [comment]
merge for the win?

**Pull request merged**